### PR TITLE
Pem 7511 azure secret cloud support in palette

### PR DESCRIPTION
**Pull request by HigashikataZhangsuke**
## Summary
This PR contains the system support for deploying clusters inside AzSecret region. 

## Jira Ticket
Linked to [PEM-7511](https://spectrocloud.atlassian.net/browse/PEM-7511)

## Changes
- See this doc for detailed info (https://docs.google.com/document/d/1bok2qKyULGFSNtTYNSp9JtFT26dhEcMpcbe3gCKqhJc/edit?usp=sharing)

## Testing Results

Here's the cluster control plane's final status after modification. All the components worked properly here.
image

[PEM-7511]: https://spectrocloud.atlassian.net/browse/PEM-7511?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by HigashikataZhangsuke** [comment]
The azure_secret_cloud.go file is the same as the PEM-7051, so it should be OK to merge.

**Comment by HigashikataZhangsuke** [comment]
Will try make a PR to upstream first.

**Comment by HigashikataZhangsuke** [comment]
Testing on **Public Cloud**:

Get E0710 07:09:53.142524       1 cluster_accessor.go:257] "Connect failed" err="error creating HTTP client and mapper: cluster is not reachable: Get \"https://tickettest-cf88cef6. │
│ centralus.cloudapp.azure.com:6443/?timeout=5s\": context deadline exceeded" controller="clustercache" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="clust │
│ er-67f99265f28ecf3ae99b0e94/tickettest" namespace="cluster-67f99265f28ecf3ae99b0e94" name="tickettest" reconcileID="fb88060c-fd29-4669-b506-3ca61466f463" 

**And eventually until the time limit of the CAPZ. When the CAPZ&CAPI deployment scale to 0, re-scale it to 1, could get the worker pool done.  However, login into the cp node, cannot find k8s under /etc/ folder.** Suspecting config errors?

**Observed the same thing for the current CAPZ version. So it should be ready to merge.**

**Comment by HigashikataZhangsuke** [comment]
For testing on the Azsecret, get same block at the palette side, could see the CAPZ side works fine. So now will focus on Palette.

File: azure/scope/azure_secret_cloud.go [hunk 1]
```diff
@@ -22,6 +43,14 @@ func init() {
 	path := os.Getenv(AzureEnvironentFolderEnvName)
 	log.Info("Path is", path)
 	if path == "" {
+		// If no environment folder is set, try fallback certificate path
+		certPath := "/home/ubuntu/combine-harbor-combined.crt"
```
Comment (guyni) [inline azure/scope/azure_secret_cloud.go]: I don't think you need to have a fallback certPath. If no environment folder is set, don't need to load any CA cert, just like what happen to the current public clouds.

**Comment by HigashikataZhangsuke** [comment]
Uploaded the latest version where removed unnecessary code of inject CAcert into Kubeconfig

**Comment by HigashikataZhangsuke** [comment]
Should be ready to merge.

File: azure/services/managedclusters/managedclusters.go [hunk 2]
```diff
@@ -142,33 +142,54 @@ func postCreateOrUpdateResourceHook(ctx context.Context, scope ManagedClusterSco
   The user needs to ensure to provide service principal with admin AAD privileges.
 */
 func reconcileKubeconfig(ctx context.Context, scope ManagedClusterScope, namespace string) (adminKubeConfigData []byte, userKubeConfigData []byte, err error) {
+	fmt.Printf("=== DEBUG: reconcileKubeconfig() called for cluster: %s ===\n", scope.ClusterName())
+	fmt.Printf("DEBUG: Namespace: %s\n", namespace)
+	fmt.Printf("DEBUG: IsAADEnabled: %v\n", scope.IsAADEnabled())
+	fmt.Printf("DEBUG: AreLocalAccountsDisabled: %v\n", scope.AreLocalAccountsDisabled())
```
Comment (vishu2498) [inline azure/services/managedclusters/managedclusters.go]: There are many debug statements in this PR. Try to have them as a log statements in either log level 0 or 1 (as required). This applies to all statements.

File: azure/services/managedclusters/managedclusters.go [hunk 3]
```diff
@@ -204,9 +225,17 @@ func getUserKubeConfigWithToken(ctx context.Context, userKubeConfigData []byte,
...
 	}
+
 	kubeconfig, err := clientcmd.Write(*config)
 	if err != nil {
 		return nil, errors.Wrap(err, "error while trying to marshal new user kubeconfig with token")
 	}
 	return kubeconfig, nil
 }
+
+func min(a, b int) int {
```
Comment (vishu2498) [inline azure/services/managedclusters/managedclusters.go]: Use the function `math.Min(a, b)` instead of this function to get the answer instantly.

File: azure/services/privatedns/link_reconciler.go [hunk 4]
```diff
@@ -51,7 +61,59 @@ func (s *Service) reconcileLinks(ctx context.Context, links []azure.ResourceSpec
...
+				continue
+			}
+			zoneRG := extractResourceGroupFromID(*zone.ID)
+			links, err := s.vnetLinkClient.ListByZone(ctx, zoneRG, *zone.Name)
+			if err != nil {
+				return managed, err
+			}
+			for _, link := range links {
+				if link.Properties != nil && link.Properties.VirtualNetwork != nil && link.Properties.VirtualNetwork.ID != nil && *link.Properties.VirtualNetwork.ID == vnetID {
+					// log.V(1).Info("Skipping vnet link creation as VNet is already linked to a zone with the same name across resource groups",
```
Comment (vishu2498) [inline azure/services/privatedns/link_reconciler.go]: Keep this log for debugging purposes.

File: azure/scope/azure_secret_cloud.go [hunk 5]
```diff
@@ -30,15 +53,182 @@ func init() {
...
+				if jsonData, err := os.ReadFile(filePath); err == nil {
+					//fmt.Printf("CAPZ: Loading Azure environment JSON file: %s\n", file.Name())
+					fmt.Printf("CAPZ: JSON file contents: %s\n", string(jsonData))
+				}
+
+				if env, err := azure.EnvironmentFromFile(filePath); err == nil {
+					azure.SetEnvironment(env.Name, env)
+					//fmt.Printf("CAPZ: Successfully loaded Azure environment: %s\n", env.Name)
+					//fmt.Printf("CAPZ: ResourceManagerEndpoint: %s\n", env.ResourceManagerEndpoint)
+					//fmt.Printf("CAPZ: ActiveDirectoryEndpoint: %s\n", env.ActiveDirectoryEndpoint)
```
Comment (vishu2498) [inline azure/scope/azure_secret_cloud.go]: Remove the commented out code if not required (for the complete PR).

File: azure/scope/cluster.go [hunk 6]
```diff
@@ -560,6 +563,13 @@ func (s *ClusterScope) VNetSpec() azure.ASOResourceSpecGetter[*asonetworkv1api20
 
 // PrivateDNSSpec returns the private dns zone spec.
 func (s *ClusterScope) PrivateDNSSpec() (zoneSpec azure.ResourceSpecGetter, linkSpec, recordSpec []azure.ResourceSpecGetter) {
+	// Check for bypass annotation
+	if s.AzureCluster.Annotations != nil {
+		if bypass, exists := s.AzureCluster.Annotations["capz.io/disable-private-dns"]; exists && bypass == "true" {
```
Comment (vishu2498) [inline azure/scope/cluster.go]: Use the annotation key as a constant.

Also, `bypass == strconv.FormatBool(true)` .

File: azure/scope/azure_secret_cloud.go [hunk 7]
```diff
@@ -2,43 +2,281 @@ package scope
...
+	"k8s.io/client-go/rest"
+	azurepkg "sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	AzureEnvironentFolderEnvName = "AZURE_ENVIRONMENT_FOLDER"
+	// Deprecated: Use ConfigMaps instead
+	AzureEnvironentFolderEnvName = "AZURE_ENVIRONMENT_FOLDER_CAPZ"
```
Comment (snehala27) [inline azure/scope/azure_secret_cloud.go]: What will happen in the cases where AZURE_ENVIRONMENT_FOLDER was used. This no more exists now

File: azure/scope/cluster.go [hunk 8]
```diff
@@ -150,24 +156,22 @@ func (s *ClusterScope) ASOOwner() client.Object {
...
-					ClusterName:      s.ClusterName(),
-					DNSName:          "",    // Set to default value
-					IsIPv6:           false, // Set to default value
-					Location:         s.Location(),
-					ExtendedLocation: s.ExtendedLocation(),
-					FailureDomains:   s.FailureDomains(),
-					AdditionalTags:   s.AdditionalTags(),
-				})
-			}
+	if s.ControlPlaneOutboundLB() != nil {
```
Comment (snehala27) [inline azure/scope/cluster.go]: Can we make sure this does not run for Public cloud?

File: azure/scope/cluster.go [hunk 9]
```diff
@@ -277,6 +282,10 @@ func (s *ClusterScope) LBSpecs() []azure.ResourceSpecGetter {
 				// or if the LB is of the type internal, save the only IP allowed for the frontend LB
 				if frontendIP.PublicIP != nil || frontendLB.Type == infrav1.Internal {
 					frontendLB.FrontendIPConfigs = []infrav1.FrontendIP{frontendIP}
+					// Log the IP configuration being used
+					if frontendLB.Type == infrav1.Internal && frontendIP.PrivateIPAddress != "" {
+						fmt.Printf("[CAPZ-DEBUG] Internal LB FrontendIP config - Name: %s, PrivateIP: %s\n", frontendIP.Name, frontendIP.PrivateIPAddress)
```
Comment (snehala27) [inline azure/scope/cluster.go]: Do we need this log or was it a debug log?

File: azure/services/managedclusters/spec_test.go [hunk 10]
```diff
@@ -429,4 +429,4 @@ func TestOIDCIssuerURLConfigMap(t *testing.T) {
 
 		g.Expect(actualOIDCIssuerConfigMapName).To(Equal("my-cluster-aso-oidc-issuer-profile"))
 	})
-}
+}
```
Comment (snehala27) [inline azure/services/managedclusters/spec_test.go]: Can we remove this diff?

File: azure/services/privatedns/link_reconciler.go [hunk 11]
```diff
@@ -51,7 +61,59 @@ func (s *Service) reconcileLinks(ctx context.Context, links []azure.ResourceSpec
 			continue
 		}
 
-		// we consider VnetLinks as managed if at least of the links is managed.
+		// Enhanced logic: check all zones with the same name across all resource groups
```
Comment (snehala27) [inline azure/services/privatedns/link_reconciler.go]: Can you added detailed comment here for what this function does?

File: controllers/azureasomanagedcontrolplane_controller.go [hunk 12]
```diff
@@ -321,6 +323,46 @@ func (r *AzureASOManagedControlPlaneReconciler) reconcileKubeconfig(ctx context.
 			a.Exec = nil
 			a.Token = token.Token
 		}
+
+		// PATCH POINT: Inject custom CA certificate data here
+		// This allows patching CA certificates for AAD-enabled clusters with local accounts disabled
```
Comment (snehala27) [inline controllers/azureasomanagedcontrolplane_controller.go:329]: Can you use proper logger for all the debug logs in here

File: controllers/azureasomanagedcontrolplane_controller.go [hunk 13]
```diff
@@ -352,6 +394,43 @@ func (r *AzureASOManagedControlPlaneReconciler) reconcileKubeconfig(ctx context.
 	return tokenExpiresIn, nil
 }
 
+// getCustomCACertificateForCluster returns custom CA certificate data for a specific cluster
+// This function leverages the same certificate that is used for Azure authentication
+// by checking the global AzSecretCertPool that gets populated during Azure client initialization
+func getCustomCACertificateForCluster(clusterName string) []byte {
```
Comment (snehala27) [inline controllers/azureasomanagedcontrolplane_controller.go]: Same as above

File: controllers/azuremanagedcontrolplane_reconciler.go [hunk 14]
```diff
@@ -130,6 +131,62 @@ func (r *azureManagedControlPlaneService) reconcileKubeconfig(ctx context.Contex
 		if len(kubeConfigData) == 0 {
 			continue
 		}
+
+		// PATCH POINT: Inject custom CA certificate data into kubeconfig before storing in secret
+		fmt.Printf("=== DEBUG: AMCP reconcileKubeconfig() - Processing kubeconfig %d (admin=0, user=1) ===\n", i)
```
Comment (snehala27) [inline controllers/azuremanagedcontrolplane_reconciler.go]: Please use standard logger with appropriate log level

File: main.go [hunk 15]
```diff
@@ -328,6 +328,12 @@ func main() {
 
 	restConfig := ctrl.GetConfigOrDie()
 	restConfig.UserAgent = "cluster-api-provider-azure-manager"
+
+	// Configure longer timeouts for cluster connections, especially for custom Azure environments
```
Comment (snehala27) [inline main.go]: Is this applicable only for non public?

File: azure/services/privatedns/link_reconciler.go (hunk 11 above)
Comment (HigashikataZhangsuke) [reply azure/services/privatedns/link_reconciler.go]: Sure. Met the problem since when I was doing tests, I accidently made a duplication there, and caused error. It's more like a corner case I would say?

File: main.go (hunk 15 above)
Comment (HigashikataZhangsuke) [reply main.go]: No, it's like a global one.

File: main.go [hunk 16]
```diff
@@ -328,6 +328,15 @@ func main() {
...
 	restConfig := ctrl.GetConfigOrDie()
 	restConfig.UserAgent = "cluster-api-provider-azure-manager"
+
+	// Configure longer timeouts for cluster connections only for Azure Secret cloud
+	if os.Getenv("AZURE_ENVIRONMENT") == azure.AzSecretCloudName {
+		restConfig.Timeout = 30 * time.Second
+	}
+
+	restConfig.QPS = 20
+	restConfig.Burst = 30
```
Comment (vishu2498) [inline main.go]: Can we move these configurations in the azure secret cloud environment only?

Comment (HigashikataZhangsuke) [reply main.go]: Sure, will do it tmr.

**Review by snehala27 — APPROVED** [review]
